core/src/main/java/org/elasticsearch/index/search/MultiMatchQuery.java

```diff
@@ -91,9 +92,12 @@ public Query parse(MultiMatchQueryBuilder.Type type, Map<String, Float> fieldNam
                 break;
             default:
                 throw new IllegalStateException("No such type: " + type);
+            }
+            final List<? extends Query> queries = queryBuilder.buildGroupedQueries(type, fieldNames, value, minimumShouldMatch);
+            result = queryBuilder.combineGrouped(queries);
```

Comment:
Yes, thats the idea. Just having one exit point IMHO is more readable and this way I can also do the assert in one place.